compiler/daml-lf-evaluator/src/DA/Daml/LF/Evaluator/Simp.hs

```diff
@@ -0,0 +1,255 @@
+    Just mainId -> do
+      runEffect mainId world $ do
+        simpExprValName mainId moduleName name >>= \case
+          Just exp -> return exp
+          Nothing -> Fail "simplify, Nothing"
+
+simpExprValName :: LF.PackageId -> LF.ModuleName -> LF.ExprValName -> Effect (Maybe Exp)
+simpExprValName pid moduleName name = do
+  getModule pid moduleName >>= \case
+    Nothing -> return Nothing
+    Just mod -> do
+      let LF.Module{moduleValues} = mod
+      case NM.lookup name moduleValues of
+        Nothing -> Fail $ "simpExprValName, " <> show name
+        Just dval -> do
+          let LF.DefValue{dvalBody=expr} = dval
+          let key = Exp.DefKey (pid,moduleName,name)
+          i <- Share key $ simpExpr expr
+          return $ Just $ Exp.Ref i
+
+getModule :: LF.PackageId -> LF.ModuleName -> Effect (Maybe LF.Module)
+getModule pid moduleName = do
+  package <- GetPackage pid
+  let LF.Package{packageModules} = package
+  case NM.lookup moduleName packageModules of
+    Nothing -> return Nothing -- Fail $ "simp-getModule, " <> show (pid,moduleName)
+    Just mod -> return $ Just mod
+
+simpExpr :: LF.Expr -> Effect Exp
+simpExpr expr = case expr of
+
+  LF.EVar name -> return $ Exp.Var name
+  LF.EVal q -> simpQualifiedExprValName q
+  LF.EBuiltin builtin -> return $ Exp.Lit $ simpBuiltin builtin
+
+  LF.ERecCon{recTypeCon=_,recFields} -> do
+    xs <- forM recFields $ \(fieldName,expr) -> do
+      e <- simpExpr expr
+      return (fieldName,e)
+    return $ Exp.Rec xs
+
+  LF.ERecProj{recTypeCon=_,recField=fieldName,recExpr} -> do
+    e <- simpExpr recExpr
+    return $ Exp.Dot e fieldName
+
+  LF.ERecUpd{} -> todo "ERecUpd"
+
+  LF.EVariantCon{varVariant=name,varArg} -> do
+    exp <- simpExpr varArg
+    return $ Exp.Con (Value.mkTag name) [exp]
+
+  LF.EEnumCon{enumDataCon=name} -> do
+    return $ Exp.Con (Value.mkTag name) []
+
+  LF.EStructCon{structFields} -> do
+    xs <- forM structFields $ \(fieldName,expr) -> do
+      e <- simpExpr expr
+      return (fieldName,e)
+    return $ Exp.Rec xs
+
+  LF.EStructProj{structField=fieldName,structExpr} -> do
+    e <- simpExpr structExpr
+    return $ Exp.Dot e fieldName
+
+  LF.EStructUpd{} -> todo "EStructUpd"
+
+  LF.ETmApp{tmappFun=func,tmappArg=arg} -> do
+    f <- simpExpr func
+    a <- simpExpr arg
+    return $ Exp.App f a
+
+  LF.ETyApp{tyappExpr=expr,tyappType=typ} -> do
+    expr <- simpExpr expr
+    return $ Exp.TypeApp expr typ
+
+  LF.ETmLam{tmlamBinder=(name,typ),tmlamBody} -> do
+    body <- simpExpr tmlamBody
+    return $ Exp.Lam name typ body
+
+  LF.ETyLam{tylamBinder=(tv,_), tylamBody=expr} -> do
+    expr <- simpExpr expr
+    return $ Exp.TypeLam tv expr
+
+  LF.ECase{casScrutinee, casAlternatives} -> do
+    scrut <- simpExpr casScrutinee
+    alts <- mapM simpAlternative casAlternatives
+    return $ Exp.Match {scrut,alts}
+
+  LF.ELet{letBinding,letBody} -> do
+    f <- simpBinding letBinding
+    body <- simpExpr letBody
+    return $ f body
```

Comment:
I worried about whether LF `let` expressions could be recursive like in Haskell. Here's what I found:

* LF expressions cannot be recursive, according to the LF spec. This isn't stated explicitly, but it's implied by the typing and evaluation rules.
* We error out on recursive let bindings during LF conversion.

So there was no reason to worry, this code is fine :-)

Reply:
That said, `let b = a + a in let c = b + b in let d = c + c in let e = d + d in ... let z = y + y in z + z` is the kind of expression that would result in an exponential blow-up with aggressive inlining. Although it is artificial.

Reply:
Actually this example *doesn't* blow up the way you suggest. Because we preserve `let`s for non duplicatable args:

    blow : Int -> Int
    blow a = let b = a + a in let c = b + b in let d = c + c in let e = d + d in e

-->

    def blow : Int64 -> Int64 =
      \(_v0 : Int64).
        let b : Int64 = ADD_INT64 _v0 _v0
            c : Int64 = ADD_INT64 b b
            d : Int64 = ADD_INT64 c c
        in ADD_INT64 d d

Reply:
For comparison, here is the original DALF:

    def blow : Int64 -> Int64 =
      \(a : Int64).
        let b : Int64 =
              532f849b7d757c5a08c76307acd720292992fcd3f89e96e660a9e2786624b957:GHC.Num:+
                @Int64
                532f849b7d757c5a08c76307acd720292992fcd3f89e96e660a9e2786624b957:GHC.Num:$fAdditiveInt
                a
                a
            c : Int64 =
              532f849b7d757c5a08c76307acd720292992fcd3f89e96e660a9e2786624b957:GHC.Num:+
                @Int64
                532f849b7d757c5a08c76307acd720292992fcd3f89e96e660a9e2786624b957:GHC.Num:$fAdditiveInt
                b
                b
            d : Int64 =
              532f849b7d757c5a08c76307acd720292992fcd3f89e96e660a9e2786624b957:GHC.Num:+
                @Int64
                532f849b7d757c5a08c76307acd720292992fcd3f89e96e660a9e2786624b957:GHC.Num:$fAdditiveInt
                c
                c
        in 532f849b7d757c5a08c76307acd720292992fcd3f89e96e660a9e2786624b957:GHC.Num:+
             @Int64
             532f849b7d757c5a08c76307acd720292992fcd3f89e96e660a9e2786624b957:GHC.Num:$fAdditiveInt
             d
             d

Reply:
Is that behavior intentional or just due to an underapproximation in `duplicatable`? I don’t understand why `ADD_INT64 b b` should not be duplicatable. It clearly has no sde effects and can be duplicated safely.

Reply:
Ah I think I misunderstood the comment there. It looks like you are trying to check if it is an atom or not to decide whether to evaluate the `let`.

Reply:
Of course, some artificial patterns will blow up: i.e.

    blow =
      let a = \x -> x+1 in
      let b = a . a in
      let c = b . b in
      let d = c . c in
      let e = d . d in
      e

Not to mention this:

    blow =
      let decrement = \x -> x-1 in
      let twice f = f . f in
      let thrice f = f . f . f in
      twice thrice thrice decrement

which will replicate the SUB operator 3^(3^2) == 19683 times :scream: